### PR TITLE
APE 14 implementation

### PR DESCRIPTION
**Pull request by nden**
APE 14 implementation.

### CODE REVIEW

**Comment by nden** [comment]
@astrofrog @Cadair This is a start of APE14 implementation. It's pretty straightforward to implement in gwcs though some questions came up, specifically the top three in the description above may need clarification.

File: gwcs/wcs.py [hunk 1]
```diff
@@ -545,3 +545,53 @@ def footprint(self, bounding_box=None, center=False):
...
+        return self.output_frame.axes_types
+    
+    def world_axis_units(self):
+        return self.output_frame.unit
+
+    def axis_correlation_matrix(self):
+        return is_separable(self.forward_transform)
+
+    def pixel_to_world_values(self, *pixel_arrays):
+        # need **kwargs
```
Comment (Cadair) [inline gwcs/wcs.py]: I am not convinced we do need this here, this API doesn't need to encapsulate the whole of the gWCS functionality.

Having said that I am not sure there is a problem with having it in the function signature. (Other than `output` should be prohibited)

Comment (nden) [reply gwcs/wcs.py]: There's a suggestion that the result should return NaNs where there's no valid WCS solution. This is controlled by keywords and their current defaults provide this functionality. So yes, currently it's not strictly needed.

**Comment by astrofrog** [comment]
@nden - just for info, I've started a document here with issues I've encountered while implementing this on the astropy side: https://docs.google.com/document/d/13bJ-tHOuI6pVmB8KtthYB_fBFIYNDF2cApJwt92G6w8/edit?usp=sharing - feel free to make suggested edits if you have any further thoughts! Once we've both had a chance to think about these issues and try and push the implementation further, my plan is to send this out to the main APE authors and then if there is agreement, to open a PR to amend the APE.

File: gwcs/wcs.py [hunk 2]
```diff
@@ -547,3 +551,98 @@ def footprint(self, bounding_box=None, center=False):
...
+        return self.output_frame.axes_type
+
+    @property
+    def world_axis_units(self):
+        return [unit.to_string() for unit in self.output_frame.unit]
+
+    @property
+    def axis_correlation_matrix(self):
+        from astropy.modeling import separable
+        return separable.is_separable(self.forward_transform)
```
Comment (Cadair) [inline gwcs/wcs.py]: I dont think this is correct, the output should be a matrix. The closes I could cook up is:
```python
s = separable.is_separable(spatial & Identity(1))
sep = np.identity(len(s), dtype=bool)
for i, a in enumerate(s):
    sep[i,i] = not(s[i])
```

Comment (nden) [reply gwcs/wcs.py]: Right. I will have to work on this. It isn't ready for review yet, especially thie property.
that said comment would be very useful.

Comment (Cadair) [reply gwcs/wcs.py]: sure, I actually came here looking for how you worked this out for compoundmodels in general and ended up getting a little distracted by matrices lol

**Comment by astrofrog** [comment]
> order of axes in pixel_shape and pixel_bounds

This is now clarified in the APE. Note that ``pixel_shape`` is now ``array_shape``

> how to pass kwargs to pixel_to_world_values and world_to_pixel_values

I don't think we allow this for now. What use case did you have in mind?

> implement world_axis_object_classes - perhaps don't require classes but a callable?

I think you can now simply implement ``world_to_pixel`` and ``pixel_to_world`` yourself and just raise a NotImplementedError for ``world_axis_object_classes`` and ``world_axis_object_components`` - this was what you were planning on doing I think?

**Comment by nden** [comment]
> >   how to pass kwargs to pixel_to_world_values and world_to_pixel_values

>I don't think we allow this for now. What use case did you have in mind?

The use case is passing non-coordinate inputs, like `spectral_order` or `slit_name`.
The general implementation of this still needs to be worked out in gwcs.
Specific examples exist in jwst.

> I think you can now simply implement world_to_pixel and pixel_to_world yourself and just raise a 
> NotImplementedError for world_axis_object_classes and world_axis_object_components - this was 
> what you were planning on doing I think?

Yes, this is fine.

**Comment by astrofrog** [comment]
@nden - is there any reason why you couldn't essentially treat ``slit_name`` or ``spectral_order`` as keys to a dictionary of WCS objects? That is: ``overall_wcs[slit_name][spectral_order].world_to_pixel(...)``. Having said that, I'm not completely opposed to the kwargs, but I just want to make sure I understand.

**Comment by nden** [comment]
@astrofrog I could but it's a dict then and not a WCS object. Many of the JWST spectral pipelines are setup as one WCS object. For example an IFU observation has one WCS for all slices. The reason is that many of the transforms are the same and are shared among slices.
Another example is the slitless spectroscopy where the models are global and are parametrized by spectral order and locaiton (IIRC).

**Comment by nden** [comment]
@astrofrog Yes, I will add `**kwargs` in this implementation. I just want to make sure that the option is available in the top level interface used in nddata (if my understanding of how this is going to work is correct).

File: gwcs/api.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        """
+        return self
+
+    def pixel_to_world(self, *pixel_arrays):
+        """
+        Convert pixel values to world coordinates.
+        """
+        if not self.forward_transform.uses_quantity:
+            kwargs = {'with_units': True}
+        return self.__call__(*pixel_arrays, **kwargs)
```
Comment (Cadair) [inline gwcs/api.py]: why the `__call__`?

File: gwcs/api.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+    def serialized_classes(self):
+        """
+        Indicates whether Python objects are given in serialized form or as
+        actual Python objects.
+        """
+        return False
+
+    def world_axis_object_classes(self):
+        raise NotImplementedError()
```
Comment (Cadair) [inline gwcs/api.py]: Is the plan to implement these to facilitate a common high level API object? (In addition to gWCS implementing the high level api?)

File: gwcs/api.py (hunk 3 above)
Comment (nden) [reply gwcs/api.py]: Yes, strictly speaking, it's not necessary.

**Comment by astrofrog** [comment]
@nden - can you try updating to the latest developer version of astropy to make sure that things work properly now?

**Comment by nden** [comment]
Last ping if anyone has time to review or intends to review it and needs more time.

**Pull request merged**